cmd/ceremony/main.go

```diff
@@ -0,0 +1,280 @@
+			return errors.New("if key-type = \"ecdsa\" then ecdsa-curve is required")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.CertificatePath == "" {
+			return errors.New("certificate-path is required")
+		}
+		if cc.IssuerPath != "" {
+			return errors.New("issuer-path is not used for root ceremonies")
+		}
+		if cc.CertificateProfile == nil {
+			return errors.New("certificate-profile is required")
+		}
+		if err := verifyProfile(cc.CertificateProfile, true); err != nil {
+			return fmt.Errorf("invalid certificate-profile: %s", err)
+		}
+	case "intermediate":
+		if cc.KeyLabel == "" {
+			return errors.New("key-label is required")
+		}
+		if cc.KeyID == "" {
+			return errors.New("key-id is required")
+		}
+		if cc.KeyType != "" {
+			return errors.New("key-type is not used for intermediate ceremonies")
+		}
+		if cc.ECDSACurve != "" {
+			return errors.New("ecdsa-curve is not used for intermediate ceremonies")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.CertificatePath == "" {
+			return errors.New("certificate-path is required")
+		}
+		if cc.IssuerPath == "" {
+			return errors.New("issuer-path is required")
+		}
+		if cc.CertificateProfile == nil {
+			return errors.New("certificate-profile is required")
+		}
+		if err := verifyProfile(cc.CertificateProfile, true); err != nil {
+			return fmt.Errorf("invalid certificate-profile: %s", err)
+		}
+	case "key":
+		if cc.KeyLabel == "" {
+			return errors.New("key-label is required")
+		}
+		if cc.KeyID != "" {
+			return errors.New("key-id is not used for key ceremonies")
+		}
+		if cc.KeyType == "" {
+			return errors.New("key-type is required")
+		}
+		if cc.KeyType != "rsa" && cc.KeyType != "ecdsa" {
+			return errors.New("key-type can only be 'rsa' or 'ecdsa'")
+		}
+		if cc.KeyType == "rsa" && cc.ECDSACurve != "" {
+			return errors.New("if key-type = \"rsa\" then ecdsa-curve is not used")
+		}
+		if cc.KeyType == "ecdsa" && cc.ECDSACurve == "" {
+			return errors.New("if key-type = \"ecdsa\" then ecdsa-curve is required")
+		}
+		if cc.PublicKeyPath == "" {
+			return errors.New("public-key-path is required")
+		}
+		if cc.IssuerPath != "" {
+			return errors.New("issuer-path is not used for key ceremonies")
+		}
+		if cc.CertificatePath != "" {
+			return errors.New("certificate-path is not used for key ceremonies")
+		}
+		if cc.CertificateProfile != nil {
+			return errors.New("certificate-profile is not used for key ceremonies")
+		}
+	}
+
+	return nil
+}
+
+func main() {
+	configPath := flag.String("config", "", "Path to ceremony configuration file")
+	flag.Parse()
+
+	if *configPath == "" {
+		log.Fatal("--config is required")
+	}
+	configBytes, err := ioutil.ReadFile(*configPath)
+	if err != nil {
+		log.Fatalf("Failed to read config file: %s", err)
+	}
+	var config ceremonyConfig
+	err = json.Unmarshal(configBytes, &config)
```

Comment:
The documentation suggests the config format is YAML, but this suggests it's still JSON.

BTW, I'd love to see this tool incorporated into the integration tests so we can have example config files that we know get exercised regularly.

Reply:
We have an issue for doing this already (#3774), but have put the work off multiple times for reasons that I cannot remember. Probably something about wanting to keep the keys/certs we use stable between integration tests run (mainly for local use?). We could just run them to generate stuff, but never actually use those things, but then we are only really testing that the tools don't hard fail, we don't learn if they _actually work_.

Reply:
> We could just run them to generate stuff, but never actually use those things, but then we are only really testing that the tools don't hard fail, we don't learn if they actually work.

That seems :+1: to me as a starting point. Maybe paired with a `softhsm2-util` invocation that checks that a new thing was stored in an HSM slot without doing too much analysis of whether it's the correct thing?